cpp/src/ffi_go/ir/encoder.cpp

```diff
@@ -0,0 +1,134 @@
+#include "encoder.h"
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <vector>
+
+#include <clp/components/core/src/ffi/encoding_methods.hpp>
+#include <clp/components/core/src/ffi/ir_stream/decoding_methods.hpp>
+
+#include <ffi_go/defs.h>
+#include <ffi_go/ir/LogTypes.hpp>
+#include <ffi_go/LogTypes.hpp>
+
+namespace ffi_go::ir {
+using namespace ffi::ir_stream;
+
+namespace {
+    /**
+     * Generic helper for ir_encoder_encode_*_log_message
+     */
+    template <class encoded_var_view_t>
+    auto encode_log_message(
+            StringView log_message,
+            void* ir_encoder,
+            StringView* logtype,
+            encoded_var_view_t* vars,
+            StringView* dict_vars,
+            Int32tSpan* dict_var_end_offsets
+    ) -> int {
+        typedef typename std::conditional<
+                std::is_same_v<Int64tSpan, encoded_var_view_t>,
+                ffi::eight_byte_encoded_variable_t,
+                ffi::four_byte_encoded_variable_t>::type encoded_var_t;
+        Encoder<encoded_var_t>* encoder{static_cast<Encoder<encoded_var_t>*>(ir_encoder)};
+        LogMessage<encoded_var_t>& ir_log_msg = encoder->m_log_message;
+        ir_log_msg.reserve(log_message.m_size);
+
+        std::string_view const log_msg_view{log_message.m_data, log_message.m_size};
+        std::vector<int32_t> dict_var_offsets;
```

Comment:
I think this vector can be a part of the ir_encoder storage. In that way, the vector can be reused to save some init/deinit and vector grow overhead.